### PR TITLE
Part pricing cache

### PR DESCRIPTION
**Pull request by SchrodingersGat**
This PR represents the first steps towards a major overhaul of pricing.

Currently, part pricing is too expensive to calculate "on the fly" and causes issues with viewing prices in a Bill of Materials, for example.

The major goal of this PR is to add a new database model which stores cached pricing information for a Part. This pricing information can be used in API requests for fast lookup against multiple parts.

Individual "quantity pricing" for a single part will still need to be calculated "on the fly" but for a single part should not be too expensive.

### TODO

- [x] Overhaul part pricing tab
- [x] Validate page load times (currently takes ~14s to load "MAST Master Assembly" page with pricing turned on)
- [x] Fix template rendering issues which clip price values to 2 decimal places
- [x] Fix BOM export issues which clip price values to 2 decimal places
- [x] Implement functionality for "caching" min and max part pricing (inc. unit tests)
- [x] Add new pricing information to various API endpoints
- [x] Display pricing information in Part table
- [x] Display pricing information in BOM table
- [x] Add background tasks for updating various pricing data
- [x] Call background tasks when underlying data changes
- [x] Add data migration to perform initial part pricing calculations
- [x] Unit testing
- [x] Periodically check for any pricing information which is stored in the "wrong" currency
- [x] Periodically check for any parts which are missing pricing information
- [ ] ~~Schedule update when part is created / edited / deleted~~
- [x] Schedule update when BomItem is is created / edited / deleted
- [x] Schedule update when SupplierPriceBreak is is created / edited / deleted
- [x] Schedule update when InternalPrice is created / edited / deleted
- [x] Schedule update when PurchaseOrderLineItem is created / edited / deleted
- [x] Updated documentation - https://github.com/inventree/inventree-docs/issues/388
- [x] Remove or refactor `has_complete_bom_pricing` method
- [x] Prevent infinite loops for updates
- [x] Handle conversion issues when rates are missing (or other errors)
- [x] Include pricing summary information when exporting BOM table
- [x] Include pricing summary information when exporting part table
- [x] Respect "internal price overrides BOM price" if set

### Future Work
- [ ] Refactor 'quantity pricing' calculations
- [ ] Refactor `get_price_info` method
- [ ] Refactor `get_bom_price_range` method
- [ ] Refactor `get_price_range` method
-----

- Closes https://github.com/inventree/InvenTree/issues/1059
- Closes https://github.com/inventree/InvenTree/issues/1656
- Closes https://github.com/inventree/InvenTree/issues/2316
- Closes https://github.com/inventree/InvenTree/issues/2891
- Closes https://github.com/inventree/InvenTree/issues/3321
- Closes https://github.com/inventree/InvenTree/issues/3326
- Closes https://github.com/inventree/InvenTree/issues/3409
- Closes https://github.com/inventree/InvenTree/issues/3460
- Closes https://github.com/inventree/InvenTree/issues/3553

### References

- https://github.com/inventree/InvenTree/issues/3409
- https://github.com/inventree/InvenTree/issues/3326
- https://github.com/inventree/InvenTree/issues/3321
- https://github.com/inventree/InvenTree/issues/1694
- https://github.com/inventree/InvenTree/issues/1989
- https://github.com/inventree/InvenTree/issues/2042
- https://github.com/inventree/InvenTree/issues/2316
- https://github.com/inventree/InvenTree/issues/2507
- https://github.com/inventree/InvenTree/issues/2891

<a href="https://gitpod.io/#https://github.com/inventree/InvenTree/pull/3710"></a>

### CODE REVIEW

File: InvenTree/part/models.py [hunk 1]
```diff
@@ -2165,6 +2181,173 @@ def after_save_part(sender, instance: Part, created, **kwargs):
...
+    Note that this pricing information does not take "quantity" into account.
+    Quantity pricing still needs to be calculated.
+    """
+
+    def save(self, *args, **kwargs):
+        """Whenever pricing model is saved, automatically update overall prices"""
+
+        self.update_overall_cost()
+
+        super().save(*args, **kwargs)
```
Comment (matmair) [inline InvenTree/part/models.py:2209]: Q: shouldn't that be saved first

**Review by matmair — COMMENTED** [review]
Seems a bit verbose datastructure wise

**Comment by SchrodingersGat** [comment]
> Seems a bit verbose datastructure wise

Which fields would you cull? How do you see the caching functionality working here?

**Comment by SchrodingersGat** [comment]
@matmair this one is actually pretty close to being ready now. There's a *lot* going on here but I'll try to briefly describe the concept.

Basically, calculating pricing "on the fly" is very expensive. For complex (multi-level) BOMs is quickly becomes prohibitive.

The goal here is to have pricing data available in real time for Bills of Material. To achieve this, pricing data gets "cached" for each part, so it can be retrieved without any need to calculate on the fly.

Now, displaying a BOM with pricing data is possible without major delay. As an example, a [part with a complex multi-level BOM](https://demo.inventree.org/part/113/) took upwards of 20 seconds to load the BOM table with pricing enabled. With price caching, this is now ~200ms.

Pricing data can come from a variety of sources:

- Internal pricing
- Purchase price history
- Supplier pricing data
- BOM pricing data

Different parts will have different pricing based on the above. Note that if enabled and available, "internal pricing" can override other pricing sources.

Pricing cache data gets updated in multiple ways:

- Manually, user interaction on the "pricing" tab
- Automatically, when underlying data changes (e.g. supplier part pricing is updated, causes price cache to be recalculated)
- Periodically, e.g. if pricing is out of date for a certain part it is recalculated

This PR also significantly overhauls the pricing tab. There's a lot of new information here, a brief example screenshot is below. A "summary" section at the top is provided, with links to detail views of each pricing data.

image

Note that this pricing tab now makes exclusive use of the API, so the page does not take ages to load due to template rendering.

Part pricing data can now also be exported in BOM files - which is very useful.

The one thing I have not yet changed is the "part quantity pricing calculation". This still requires on-the-fly calculations to determine quantity pricing (e.g. for complex BOMs) so the caching does not solve it yet. Perhaps a future update can improve that.

Let me know what you think. I am hoping it is polished enough to make it into 0.9.0, I certainly need this for my work.

**Comment by matmair** [comment]
@SchrodingersGat that is a huge one. I think we will need several blog entries to go into depth on all the cool stuff between 0.8.x and 0.9.x
I would like to take some time reviewing this - is it ok if I take a few days to test it out and get some real world performance data?

**Comment by SchrodingersGat** [comment]
Absolutely take your time, we want to get this right. I only have a few little window dressing changes to make now. And docs.

**Review by matmair — APPROVED** [review]
@SchrodingersGat looks good from my side.
Still wondering if code duplication could be reduced with the different types of pricing (esp. in API urls, forms)

**Comment by matmair** [comment]
Not really sure if it solves https://github.com/inventree/InvenTree/issues/3553 fully

**Comment by SchrodingersGat** [comment]
> Not really sure if it solves https://github.com/inventree/InvenTree/issues/3553 fully

@matmair it extends all pricing fields to use 6 decimal places, is this not fixing that?

**Comment by matmair** [comment]
@SchrodingersGat i think the idea was originally to make a toggle as we also had an FR to reduce the number of shown decimal places.

**Pull request merged**